flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/FileLock.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.testutils;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.file.Paths;
+
+/** A file lock used for avoiding race condition among multiple unit test processes */
+public class FileLock {
+    private final String fileName;
+    private FileOutputStream outputStream;
+    private java.nio.channels.FileLock lock;
+
+    public FileLock(String fileName) throws Exception {
+        this.fileName = fileName;
```

Comment:
The lock is currently locking on the test name. However, the resource under contention is actually the port number.
So I propose that `NetUtils#getAvailablePort` determines a free port and if it finds any, it adds a lock to it (using some new `tryLock`). If that doesn't work, `getAvailablePort` just tries the next free port and repeats.

To unlock the port, you probably need to add a first-class object `Port` or `RandomPort` that implements `AutoClosable` and which then releases the owned file lock.